client-apis/encrypt.md

```diff
@@ -67,24 +72,31 @@ A bound on the length of the [plaintext](#plaintext) to encrypt.
 This behavior MUST output the following if the behavior is successful:
 
 - [message](#message.md)
-
-This behavior MAY output the following:
-
 - [keyring trace](#structures.md#keyring-trace)
 
 To construct the outputs, some fields MUST be constructed using information obtained
 from a set of valid [encryption materials](#structures.md#encryption-materials).
 This behavior MUST obtain this set of [encryption materials](#structures.md#encryption-materials)
 by calling [Get Encryption Materials](#cmm-interface.md#get-encryption-materials) on a [CMM](#cmm-interface.md).
+
 The CMM used MUST be the input CMM, if supplied.
 If instead the user supplied a [keyring](#keyring-interface.md), this behavior MUST use a [default CMM](#default-cmm.md),
 constructed using the user supplied keyring as input.
-The call to [Get Encryption Materials](#cmm-interface.md#get-encryption-materials) MUST include the
-input [encryption context](#encryption-context), if supplied.
-If the length is known on the input [plaintext](#plaintext), this call MUST also include that value
-
-The [algorithm suite](#algorithm-suites.md) used in all aspects of this behavior MUST be the algorithm suite in the
+The call to [Get Encryption Materials](#cmm-interface.md#get-encryption-materials) on that CMM
+MUST be constructed as follows:
+
+- Encryption Context: If provided, this is the [input encryption context](#encryption-context).
+  Otherwise, this is an empty encryption context.
+- Algorithm Suite: If provided, this is the [input algorithm suite](#algorithm-suite).
+  Otherwise, this field is not included.
+- Plaintext Length: If provided, this is the [input plaintext length](#plaintext-length).
+  Otherwise, if the length is known on the input [plaintext](#plaintext), this value MUST equal to
```

Comment:
What happens if you know the length _and_ the provided value is smaller?

Reply:
The API should honor the API contract first, so it should go with the smaller value, which means that it is the responsibility of the encrypt API to not encrypt more than that.

For the scope of this PR, this means that when getting encryption materials the provided value should be used, which should be what the language above says to do.

Reply:
I disagree.
I think...

Say we try and encrypt the string `secret`
and pass a length of `3`.
It is unacceptable to return an encrypted payload that will decrypt to `sec`.

I say we must error is we are unable to encrypt the message.
In the ESDK-JS,
if you pass plaintext with a knowable length,
I will clobber whatever optional length with the known value
https://github.com/aws/aws-encryption-sdk-javascript/blob/master/modules/encrypt-node/src/encrypt.ts#L31
In the case of an unknown length,
the plaintext length is the maximum
https://github.com/aws/aws-encryption-sdk-javascript/blob/master/modules/encrypt-node/src/framed_encrypt_stream.ts#L73

Reply:
After talking offline:

1. We're in agreement that the example you give above showcases a behavior we want to make sure our API MUST NOT do.
2. We agree that if the API will always know the length of the plaintext (ie exclusively a one shot API), then the API MUST NOT give the user the option to specify a plaintext length.
3. In the case that it is possible for the plaintext length to be unknown, then the API MUST have an OPTIONAL input to specify the plaintext length.
3.5. In this case, if during the Encrypt operation, it turns out the plaintext length *is* known, and the user specified a plaintext length less than the actual plaintext length, Encrypt MUST NOT do the behavior described above.

So open question: What behavior MUST it do instead? As you said, JS clobbers the value. My preference would be to fail (in whatever way makes sense for the implementation).

Second question: What should be done if the supplied plaintext length is greater than the actual length? Should we return an error, should we use the actual length, or should we use the supplied length? All current implementations do the third, and I would argue that that allows users to depend on weird logic that is not the intended purpose of the plaintext length field.

Thoughts?

Reply:
cutting specifying this from the scope of this PR. This will be addressed in #138.